docs: add request for maintainership md file (#2453)
cosmos/ibc-go: MAINTAINERSHIP.md

# Request For Maintainership

> This document details the acceptance process for requests from external contributors who wish for the `ibc-go` core team to take over **maintainership** of a complex block of code (ie: an application module or a light client implementation). It is the process we also follow internally for features that we develop which will go into the `ibc-go` codebase.

Reviewer: ```suggestion
This document details the acceptance process for requests from external contributors who require for the `ibc-go` core team to take over **maintainership** of a complex block of code (ie: an application module or a light client implementation). It is the process we also follow internally for features that we develop which will go into the `ibc-go` codebase.
```

Reviewer: or different wording. I think there should be a stronger requirement than desiring for someone else to take over responsibility for a piece of code. We should encourage the community not to be reliant on us

> For projects who have created a custom IBC application and want us to list this application on the registry, please break out your module into its own repo for ease of import into `ibc-go`!

Reviewer: > or ease of import into `ibc-go`

For ease of important into chains?

Author: no, i mean, that each feature that we develop for ibc-go follows this process of development

- Create a repo with the custom module in a folder `x/` or `modules/`. 
- Please include an app that contains the custom module along with end-to-end tests that spins up the blockchain and runs the custom module. 
- See [cosmos/interchain-security](https://github.com/cosmos/interchain-security) for an example of this setup.

For contributors wishing to submit contributions to the codebase, please check our [Contributor Guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md) :)

<p align="center">

>   <img src="maintainership.png?raw=true" alt="maintainership" width="80%" />

Reviewer: One question about the image, if there isn't an ICS spec, there should be right? Or is it supposed to imply an ICS spec won't always be necessary?

Author: some cases (specified in the spec section below) will not require an ics spec

Reviewer: "Is there an ICS spec?" I think can be "Is an ICS spec required?" ?

</p>

## Step 1: Product check

Reach out to the IBC product team through susannah@interchain.io to coordinate use-case walkthrough.

> Answer these questions in a requirements doc:

Reviewer: In addition, there should be a need for ibc-go maintainers to take over this code. We should encourage the community not to be reliant on us

```
    What problem does this feature solve?

    What are the current solutions or workarounds? 

    Are there other versions or implementations?

    If there are other versions of this feature, why is this solution better?

    What are the use cases?

    Which users have confirmed they will use this? 

    How urgent is it to implement this feature?

    How soon after being developed would this be adopted? 

    What is the impact of this feature being adopted?
```

Answers to these questions should also be detailed in an issue in the `ibc-go` repo to open up the discussion to a wider audience, this can be done before or after the walk through. 

> The acceptance criteria is based on the answers to these questions - the feature must solve a genuine problem, have users that would greatly benefit from the solution and be generic enough to benefit many users of the `ibc-go` implementation. 

Reviewer: I don't think the product check should be the final consideration for acceptance into the codebase. There should be some code/spec considerations (though minimal). Mostly want to protect against, unfinished/untested/undocumented code or poorly designed specifications

## Step 2: Submit spec to the IBC protocol repo

> A detailed review of the specification can be expected **within 2 weeks** of submission. Please notify the specification team if this does not occur, so it can be corrected as soon as possible.

Reviewer: Is it worth limiting expectations somewhat here? At the moment this seems like an expectation for @AdityaSripal to review the spec within 2 weeks. What if he is on vacation? Maybe it is better to say the submitter will be given an expected timeline for specification review. I agree it is important to commit to some expected turnaround time

The product check should happen before the spec review and this states "within 2 weeks of submission" which may not give much time for spec review

Author: within 2 weeks of submission means 2 weeks of submission for the spec, not the product check, i can clarify that. re: timeline... i can add that this will be communicated explicitly but expected timeline is 2 weeks. how does that sound?

Any IBC code that is expected to be implemented across different chains in order to function correctly must be specified in the IBC repo to be accepted unless exempted by the specification team. 
    

> Unilateral software (ie. code that only needs to run on a single chain to be functional) need not be submitted.

Reviewer: There should still be some sort of design document in my opinion. Maybe link ADR 008 as an example?

If the associated module to be developed is expected to be submitted to the `ibc-go` team for maintainership, this should already be flagged at this step so that we can start thinking about/preparing our own capacity for the engineering team.

## Step 3: Prepare code for handover

*(this step can be initiated in parallel w/ spec submission, subject to feature complexity)*

> Once the spec has been given initial approval, `ibc-go` engineering will coordinate a code walkthrough in preparation for taking the module into the repo.

Reviewer: I think it can be expected we will have requested changes after the code walkthrough (we have for our own code walkthroughs). The developers should be prepared to address these changes in a timely manner

The code that is presented should adhere to our [code contributor guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md).

> More details on what code walkthrough should cover will be provided by the `ibc-go` engineering team on a case by case basis. However, the code should at least be unit and E2E tested. Think about preparing for this process similarly to submitting a codebase for audit :)

Reviewer: The code walkthrough should cover every line of the codebase in my opinion. We should understand why each line was written

> More details on what code walkthrough should cover will be provided by the `ibc-go` engineering team on a case by case basis. However, the code should at least be unit and E2E tested. Think about preparing for this process similarly to submitting a codebase for audit :)

Reviewer: ```suggestion
More details on what code walkthrough should cover will be provided by the `ibc-go` engineering team on a case by case basis. However, the code should sufficiently unit and E2E tested. Think about preparing for this process similarly to submitting a codebase for audit :)
```

For projects who have created a custom IBC application and want us to list this application on the registry, please break out your module into its own repo for ease of import into chains!

- Create a repo with the custom module in a folder `x/` or `modules/`. 
- Please include an app that contains the custom module along with end-to-end tests that spins up the blockchain and runs the custom module. 
- See [cosmos/interchain-security](https://github.com/cosmos/interchain-security) for an example of this setup.

For contributors wishing to submit contributions to the codebase, please check our [Contributor Guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md) :)

<p align="center">
  <img src="maintainership.png?raw=true" alt="maintainership" width="80%" />
</p>

## Step 1: Product check

Reach out to the IBC product team through susannah@interchain.io to coordinate use-case walkthrough.

Answer these questions in a requirements doc:

> ```

Reviewer 2: I would just write this in normal text as an unordered list.

    Are there other versions or implementations?

    If there are other versions of this feature, why is this solution better?

    What are the use cases?

    Which users have confirmed they will use this? 

    How urgent is it to implement this feature?

    How soon after being developed would this be adopted? 

    What is the impact of this feature being adopted?

    Is there a specific need for this feature to be included in the `ibc-go` codebase, rather than in its own module repo?
```

Answers to these questions should also be detailed in a **discussion** in the `ibc-go` repo to open up the discussion to a wider audience, this can be done before or after the walk through. 

The acceptance criteria is based on the answers to these questions and the results of this product check, as well as of course an acceptable spec should the module be deemed to need one. Please see Step #2 below for the spec considerations.

In summary, the feature must solve a genuine problem, have users that would greatly benefit from the solution and be generic enough to benefit many users of the `ibc-go` implementation. 

## Step 2: Submit spec to the IBC protocol repo

> A detailed review of the specification can be expected **within 2 weeks** of submission of the specification to the repo. Please notify the specification team if this does not occur, so it can be corrected as soon as possible. Please note that this timeline may be subject to amendment based on complexity of the spec and team capacity considering other ongoing reviews, but we will strive to ensure a 2 week turnaround.

Reviewer 3: should we include an email address for the `specification team` as we need for product?

Author: good question. @AdityaSripal do you have a preference here?

Any IBC code that is expected to be implemented across different chains in order to function correctly must be specified in the IBC repo to be accepted unless exempted by the specification team. 
    
Unilateral software (ie. code that only needs to run on a single chain to be functional) need not be submitted. In these cases however, some sort of design document such as [ADR-008](https://github.com/cosmos/ibc-go/pull/1976/files) should be submitted.

If the associated module to be developed is expected to be submitted to the `ibc-go` team for maintainership, this should already be flagged at this step so that we can start thinking about/preparing our own capacity for the engineering team.

## Step 3: Prepare code for handover

*(this step can be initiated in parallel w/ spec submission, subject to feature complexity)*

> Once the spec has been given initial approval, `ibc-go` engineering will coordinate a code walkthrough in preparation for taking the module into the repo. Any requested changes from the `ibc-go` engineering team after the code walkthrough should be discussed and/or addressd in a timely manner.

Reviewer 3: ```suggestion
Once the spec has been given initial approval, `ibc-go` engineering will coordinate a code walkthrough in preparation for taking the module into the repo. Any requested changes from the `ibc-go` engineering team after the code walkthrough should be discussed and/or addressed in a timely manner.
```

The code that is presented should adhere to our [code contributor guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md).

> More details on what code walkthrough should cover will be provided by the `ibc-go` engineering team on a case by case basis. However, the code should be sufficiently unit and E2E tested. Think about preparing for this process similarly to submitting a codebase for audit :)

Reviewer 3: [nit] we are using `walkthrough` and `walk through`, we should use one to stay consistent. 
[mega-nit] we should keep punctuation consistent (missing full stop)

Reviewer 3: Maybe we can link to the [E2E Readme](https://github.com/cosmos/ibc-go/blob/main/e2e/README.md#how-to-write-tests) here

> Please indicate the expected contribution of your team maintainership, if any. This contribution should also include ideas about devrels support and support for product on social media

Reviewer 3: ```suggestion
Please indicate the expected contribution of your team maintainership, if any. This contribution should also include ideas about devrels support and support for product on social media.
```
